### PR TITLE
cmd/tailscale/web: support TLS from env vars.

### PR DESCRIPTION
**Pull request by DentonGentry**
pfSense stores its SSL certificate and key in the PHP config.
We wrote PHP code to pull the two out of the PHP config and
into environment variables before running "tailscale web".

The pfSense web UI is served over https, we need "tailscale web"
to also support https in order to put it in an <iframe>.

Signed-off-by: Denton Gentry <dgentry@tailscale.com>

### CODE REVIEW

File: cmd/tailscale/cli/web.go [hunk 1]
```diff
@@ -83,6 +85,35 @@ var webArgs struct {
 	cgi    bool
 }
 
+func tlsCertFromEnvironment(listen string) (*http.Server, error) {
+	crt := os.Getenv("TLS_CRT")
+	key := os.Getenv("TLS_KEY")
+	if crt != "" && key != "" {
+		fmt.Println(crt)
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:92]: debug? delete?

File: cmd/tailscale/cli/web.go [hunk 2]
```diff
@@ -83,6 +85,35 @@ var webArgs struct {
 	cgi    bool
 }
 
+func tlsCertFromEnvironment(listen string) (*http.Server, error) {
+	crt := os.Getenv("TLS_CRT")
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:89]: these names would make me assume these are paths to files on disks, not the PEM contents themselves.

File: cmd/tailscale/cli/web.go [hunk 3]
```diff
@@ -83,6 +85,35 @@ var webArgs struct {
 	cgi    bool
 }
 
+func tlsCertFromEnvironment(listen string) (*http.Server, error) {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:88]: it's named as if it returns a "tlsCert" but it returns a Server. And it returning (nil, nil) is worth documenting.

File: cmd/tailscale/cli/web.go [hunk 4]
```diff
@@ -83,6 +85,35 @@ var webArgs struct {
...
+		cert, err := tls.X509KeyPair([]byte(crt), []byte(key))
+		if err != nil {
+			return nil, err
+		}
+
+		tlsConfig := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		}
+
+		// Build a server:
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:105]: comment doesn't add anything. the server and tlsConfig variables don't much either. could just `return &http.Server{...}, nil` probably

File: cmd/tailscale/cli/web.go [hunk 5]
```diff
@@ -83,6 +85,35 @@ var webArgs struct {
 	cgi    bool
 }
 
+func tlsCertFromEnvironment(listen string) (*http.Server, error) {
+	crt := os.Getenv("TLS_CRT")
+	key := os.Getenv("TLS_KEY")
+	if crt != "" && key != "" {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:91]: Go in general prefers early returns and outdenting

so inverting this,

```go
   if crt == "" || key == "" {
       return nil, nil
   }
   ....
```

File: cmd/tailscale/cli/web.go [hunk 6]
```diff
@@ -96,8 +127,14 @@ func runWeb(ctx context.Context, args []string) error {
 		return nil
 	}
 
-	log.Printf("web server running on: %s", urlOfListenAddr(webArgs.listen))
-	return http.ListenAndServe(webArgs.listen, http.HandlerFunc(webHandler))
+	server, err := tlsCertFromEnvironment(webArgs.listen)
+	if server != nil && err == nil {
```
Comment (bradfitz) [inline cmd/tailscale/cli/web.go:131]: You don't handle the err != nil case.

File: cmd/tailscale/cli/web.go (hunk 1 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: Removed.

File: cmd/tailscale/cli/web.go (hunk 2 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: Added "_PEM" to environment variable names.

File: cmd/tailscale/cli/web.go (hunk 3 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: Renamed tlsConfigFromEnvironment and made it return tls.Config.

File: cmd/tailscale/cli/web.go (hunk 4 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: Made it return tls.Config.

File: cmd/tailscale/cli/web.go (hunk 5 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: Done.

File: cmd/tailscale/cli/web.go (hunk 6 above)
Comment (DentonGentry) [reply cmd/tailscale/cli/web.go]: No longer has the err here.

**Review by bradfitz — APPROVED** [review]

**Pull request merged**